internal/incremental/snapshot.go

```diff
@@ -0,0 +1,569 @@
+	latestChangedDtsFile string
+	// Hash of d.ts emitted for the file, use to track when emit of d.ts changes
+	emitSignatures map[tspath.Path]*emitSignature
+	// Recorded if program had errors
+	hasErrors core.Tristate
+	// If semantic diagnsotic check is pending
+	checkPending bool
+
+	// Additional fields that are not serialized but needed to track state
+
+	// true if build info emit is pending
+	buildInfoEmitPending                    bool
+	hasErrorsFromOldState                   core.Tristate
+	allFilesExcludingDefaultLibraryFileOnce sync.Once
+	//  Cache of all files excluding default library file for the current program
+	allFilesExcludingDefaultLibraryFile []*ast.SourceFile
+
+	// Used with testing to add text of hash for better comparison
+	hashWithText bool
+}
+
+func (s *snapshot) createEmitSignaturesMap() {
+	if s.emitSignatures == nil && s.options.Composite.IsTrue() {
+		s.emitSignatures = make(map[tspath.Path]*emitSignature)
+	}
+}
+
+func (s *snapshot) addFileToChangeSet(filePath tspath.Path) {
+	s.changedFilesSet.Add(filePath)
+	s.buildInfoEmitPending = true
+}
+
+func (s *snapshot) addFileToAffectedFilesPendingEmit(filePath tspath.Path, emitKind FileEmitKind) {
+	existingKind := s.affectedFilesPendingEmit[filePath]
+	if s.affectedFilesPendingEmit == nil {
+		s.affectedFilesPendingEmit = make(map[tspath.Path]FileEmitKind)
+	}
+	s.affectedFilesPendingEmit[filePath] = existingKind | emitKind
+	delete(s.emitDiagnosticsPerFile, filePath)
+	s.buildInfoEmitPending = true
+}
+
+func (s *snapshot) getAllFilesExcludingDefaultLibraryFile(program *compiler.Program, firstSourceFile *ast.SourceFile) []*ast.SourceFile {
+	s.allFilesExcludingDefaultLibraryFileOnce.Do(func() {
+		files := program.GetSourceFiles()
+		s.allFilesExcludingDefaultLibraryFile = make([]*ast.SourceFile, 0, len(files))
+		addSourceFile := func(file *ast.SourceFile) {
+			if !program.IsSourceFileDefaultLibrary(file.Path()) {
+				s.allFilesExcludingDefaultLibraryFile = append(s.allFilesExcludingDefaultLibraryFile, file)
+			}
+		}
+		if firstSourceFile != nil {
+			addSourceFile(firstSourceFile)
+		}
+		for _, file := range files {
+			if file != firstSourceFile {
+				addSourceFile(file)
+			}
+		}
+	})
+	return s.allFilesExcludingDefaultLibraryFile
+}
+
+func getTextHandlingSourceMapForSignature(text string, data *compiler.WriteFileData) string {
+	if data.SourceMapUrlPos != -1 {
+		return text[:data.SourceMapUrlPos]
+	}
+	return text
+}
+
+func (s *snapshot) computeSignatureWithDiagnostics(file *ast.SourceFile, text string, data *compiler.WriteFileData) string {
+	var builder strings.Builder
+	builder.WriteString(getTextHandlingSourceMapForSignature(text, data))
+	for _, diag := range data.Diagnostics {
+		diagnosticToStringBuilder(diag, file, &builder)
+	}
+	return s.computeHash(builder.String())
+}
+
+func diagnosticToStringBuilder(diagnostic *ast.Diagnostic, file *ast.SourceFile, builder *strings.Builder) string {
```

Comment:
```suggestion
func diagnosticToStringBuilder(diagnostic *ast.Diagnostic, file *ast.SourceFile, builder *strings.Builder) {
```

Reply:
Just the edit isn't enough, since the returns are still present